selfservice/strategy/oidc/provider_x.go

```diff
@@ -0,0 +1,164 @@
+// Copyright © 2024 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package oidc
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/dghubble/oauth1"
+	"github.com/dghubble/oauth1/twitter"
+	"github.com/pkg/errors"
+
+	"github.com/ory/herodot"
+)
+
+var _ Provider = (*ProviderX)(nil)
+var _ OAuth1Provider = (*ProviderX)(nil)
+
+const xUserInfoBase = "https://api.twitter.com/1.1/account/verify_credentials.json"
+const xUserInfoWithEmail = xUserInfoBase + "?include_email=true"
+
+type ProviderX struct {
+	config *Configuration
+	reg    Dependencies
+}
+
+func (p *ProviderX) Config() *Configuration {
+	return p.config
+}
+
+func NewProviderX(
+	config *Configuration,
+	reg Dependencies) Provider {
+	return &ProviderX{
+		config: config,
+		reg:    reg,
+	}
+}
+
+func (p *ProviderX) ExchangeToken(ctx context.Context, req *http.Request) (*oauth1.Token, error) {
+	requestToken, verifier, err := oauth1.ParseAuthorizationCallback(req)
+	if err != nil {
+		return nil, err
+	}
+
+	accessToken, accessSecret, err := p.OAuth1(ctx).AccessToken(requestToken, "", verifier)
+	if err != nil {
+		return nil, err
+	}
+
+	return oauth1.NewToken(accessToken, accessSecret), nil
+}
+
+func (p *ProviderX) AuthURL(ctx context.Context, state string) (string, error) {
+	c := p.OAuth1(ctx)
+
+	// We need to cheat so that callback validates on return
+	c.CallbackURL = c.CallbackURL + fmt.Sprintf("?state=%s&code=unused", state)
```

Comment:
What exactly is the cheat referred to here? Do we need to set a code, but it gets ignored by twitter? Will it still be in the URL when being redirected back?

Reply:
Good question - I took over some of the code from PR #517 where this originally was at. I believe that OAuth1 doesn't return a code parameter like OAuth2, and that would break the provider logic. So the added this workaround? I'm not sure to be honest.